core/src/main/java/io/grpc/Server.java

```diff
@@ -38,6 +39,16 @@
    */
   public abstract Server start() throws IOException;
 
+  /**
+   * Returns the address the server is listening on.  This can return null if there is no actual
+   * address or the result otherwise does not make sense.  Result is undefined after the server is
+   * terminated.
+   *
+   * @throws IllegalStateException if the server has not yet been started.
+   * @since 1.0.0
+   */
+  public abstract InetSocketAddress mainAddress();
```

Comment:
gRPC also supports non TCP sockets (such as unix domain sockets), which would not be instances of `InetSocketAddress`.  Additionally, gRPC (will) support multiple listen addresses so there may not be one address.  You might consider making this:

```java
public List<SocketAddress> listeningAddresses() {
  throw new UnsupportedOperationException();
}
```

Then subclasses could override it appropriately.  Additionally, this would nicely fit with InprocessServerBuilder, as it could return an empty list.

Reply:
Carl's suggested API looks good to me as well. InProcessServerBuilder could actually return the InProcessSocketAddress.